predicators/nsrt_learning/strips_learning/effect_search_learner.py

```diff
@@ -21,160 +21,121 @@
-        }
-        new_param_option_to_groups[param_option].remove(
-            (option_spec, add_effects, keep_effects))
-        return _EffectSets(new_param_option_to_groups)
+def _update_pnad_seg_to_keep_effs(pnad: PartialNSRTAndDatastore,
+                                  necessary_image: Set[GroundAtom],
+                                  ground_op: _GroundSTRIPSOperator,
+                                  obj_to_var: Dict[Object, Variable],
+                                  segment: Segment) -> None:
+    """Updates the pnad's seg_to_keep_effs_sub dictionary, which is necesssary
+    for correctly grounding keep effects to data."""
+    # Every atom in the necessary_image that wasn't in the
+    # ground_op's add effects is a possible keep effect. This
+    # may add new variables, whose mappings for this segment
+    # we keep track of in the seg_to_keep_effects_sub dict.
+    for atom in necessary_image - ground_op.add_effects:
+        keep_eff_sub = {}
+        for obj in atom.objects:
+            if obj in obj_to_var:
+                continue
+            new_var = utils.create_new_variables([obj.type],
+                                                 obj_to_var.values())[0]
+            obj_to_var[obj] = new_var
+            keep_eff_sub[new_var] = obj
+        pnad.poss_keep_effects.add(atom.lift(obj_to_var))
+        if segment not in pnad.seg_to_keep_effects_sub:
+            pnad.seg_to_keep_effects_sub[segment] = {}
+        pnad.seg_to_keep_effects_sub[segment].update(keep_eff_sub)
 
 
 class _EffectSearchOperator(abc.ABC):
     """An operator that proposes successor sets of effect sets."""
 
     def __init__(
-            self, trajectories: List[LowLevelTrajectory],
-            train_tasks: List[Task], predicates: Set[Predicate],
-            segmented_trajs: List[List[Segment]],
-            effect_sets_to_pnads: Callable[[_EffectSets],
-                                           List[PartialNSRTAndDatastore]],
-            backchain: Callable[[
-                List[Segment], List[PartialNSRTAndDatastore], Set[GroundAtom]
-            ], _Chain], associated_heuristic: _EffectSearchHeuristic) -> None:
+        self, trajectories: List[LowLevelTrajectory], train_tasks: List[Task],
+        predicates: Set[Predicate], segmented_trajs: List[List[Segment]],
+        backchain: Callable[
+            [List[Segment], List[PartialNSRTAndDatastore], Set[GroundAtom]],
+            _Chain], associated_heuristic: _EffectSearchHeuristic,
+        spawn_new_pnad: Callable[[Segment], PartialNSRTAndDatastore],
+        get_pnads_with_keep_effects: Callable[[PartialNSRTAndDatastore],
+                                              Set[PartialNSRTAndDatastore]],
+        recompute_pnads_from_effects: Callable[[List[PartialNSRTAndDatastore]],
+                                               List[PartialNSRTAndDatastore]],
+        _clear_unnecessary_keep_effs_sub: Callable[[PartialNSRTAndDatastore],
```

Comment:
i'm thinking we should just pass in the `EffectSearchSTRIPSLearner` instance here instead of all these methods. the downside of doing that is it's less explicit what methods this operator relies on. but this signature is so complicated now that I don't think we're getting any readability benefits anyway. (admittedly, in my original implementation, we were already pushing the boundary)